### ISSUE TITLE
Return additional fields for 'multiple' pages

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero#1479]
We'd like to improve the results we show for `multiple` pages in various places (including in translation-server). Unfortunately, to do this we'll have to update translators to scrape/extract additional data, though we can do so gradually.

Basically, instead of passing something like this to `Zotero.selectItems`,

```
{
  "http://example.com/items/1": "Foo"
}
```

…I think we'd pass a simplified variation of translator item JSON:

```
{
  "http://example.com/items/1": {
    "itemType": "newspaperArticle",
    "title": "Foo",
    "creators": "J Smith, F Bar",
    "date": "2018",
    "publication": "New York Times"
  }
}
```

(We don't want to have to deal with parsing creators, so that one would just be a string.)

Everything except the title would continue to be optional, and any missing data just wouldn't be displayed. (We'd want to make sure that all attempts to extract these additional fields were allowed to fail.) The old format with just titles would continue to be supported, so we could update translators one by one.

@adam3smith, @zuphilip, any thoughts?

**Comment by adam3smith** [comment]
A couple of thoughts:

1. This is going to create more work -- a good amount of it -- for the multiple part of translators, as that data is rarely going to be available in multiple pages in a structured form. What's the reason for that? Just disambiguating search results or anything else? It's certainly doable, I'm just wondering if it's the best use of our time and also it'd help to understand what's this for to see how strongly we'd enforce trying to parse as much as possible for new translators.

2. I'd just want to flag that search results/item pages are often pretty sparse, so I think there's be a significant number of translators where we'll stick to title

3. On the plus side, if we could easily pass that information on to the `scrape` function, that would in some cases (especially where metadata on item pages are poor) make things easier.

**Comment by dstillman** [comment]
Yes, it's for disambiguating results — primarily in non–Zotero client contexts. Basically, we're adding a text-based search option (title, creator, year) to translation-server, which might result in multiple results with the same or similar titles but different authors, and we need to be able to pass those back in the result. translation-server uses the same mechanism as `selectItems` for its multi-item responses.

In Zotero you'll already be on the search page, so doing this for most translators won't be that important. This is mostly for use in search translators used in other contexts that are returning results for arbitrary queries. That said, on sites where we can do this easily, I think it'd create a nicer experience in Zotero as well.

> that data is rarely going to be available in multiple pages in a structured form

OK, that's helpful. So maybe a better plan here would be to just return `title`, `itemType`, and `description`. E.g., on Google Scholar (which is irrelevant for translation-server because it gets blocked, but still) we'd just pass through `.gs_a`. Other than inconsistency between sites, which isn't that big a deal, the main downside would be that, for services where the metadata was available in a structured format (e.g., a search API), you'd have to concatenate the fields in the translator, and it'd be a bit annoying to have to think about formatting those. So maybe it should support `title` and `itemType` and then either `creators`/`date`/`publication` or `description`, which would be understood to be mutually exclusive, so that we could at least present any structured metadata in a consistent manner.

**Comment by adam3smith** [comment]
let me survey a couple of sites and see how this looks. We're most likely to find something structured on academic sites, so I think it makes sense to look through some examples more systematically before making a decision on the data model.

**Comment by zuphilip** [comment]
I still have some difficulties to understand the use case behind this:

> This is mostly for use in search translators used in other contexts that are returning results for arbitrary queries.

There are currently search translators for doi, ISBN, PMID which are e.g. Crossref, LoC, PubMed. Do you want to search these databases with generic queries not necessarily including the identifier? What would be an realistic example of such a query?

> Yes, it's for disambiguating results — primarily in non–Zotero client contexts.

Would the disambiguation task been done manually or algorithmically?

---

The suggested changes sound like a huge amount of work! The first example in the new form would have to look like:
```
{
  "http://example.com/items/1": {
      "title": "Foo"
   }
}
```
which is different from what we have currently and therefore every translator has to be migrated. But the new migrated translators will then also only work with a new migrated Zotero resp. translation-server. Anyone who is running an older version of Zotero with the newest translators will become a problem. Maybe also problems with depending third-party software such as Citoid or [PME](https://github.com/proquest/PME)...

---

Would it also be possible to make *only* some changes for the search translators you are interested in your use case? Or using another function `Zotero.selectItemsExtended` in these translators?

**Comment by dstillman** [comment]
> Do you want to search these databases with generic queries not necessarily including the identifier?

Yes. We're extending the search translators to take a`query` parameter as an alternative to `ISBN`, `DOI`, etc.

> Would the disambiguation task been done manually or algorithmically?

It's primarily for users to be able to tell the difference between, say, two entries with the same title — again, initially for search results from translation-server, but potentially in the Select Items dialog in the Zotero Connector.

> every translator has to be migrated

No, not at all. As I said above, the old format with just titles would continue to be supported. Base support for this in any translation-server client (e.g., Citoid) is literally just `val.title || val`. And with a few extra lines, something like Citoid could show clearer results for `multiple` pages (or the new search capabilities, if they end up wanting to take advantage of those).

I really don't want to give the impression that we're asking for more work from you and other translator authors. We're planning to rewrite the Select Items dialog in the Connector at some point, at which point I think it'd make for a better user experience to present some additional details sometimes, but only when it's easy to grab a description string along with the title.

> Anyone who is running an older version of Zotero with the newest translators will become a problem.

We'll initially return this only for the new `query` searches in the search translators, which will only be used in translation-server for now, and we'll add support for the new syntax to the translation architecture. Once we cut off Zotero 4.0 translator updates (along with 4.0 syncing) in a few months (exact date TBD), we can start returning these from web translators to the extent that we want to.

**Comment by zuphilip** [comment]
Thank you @dstillman for the clarification. :relieved: It is good to hear that most translators will stay with the simple `items` syntax and only a few (e.g. search translators) will have an more elaborate `items` syntax for multiples.

I think that for disambiguation also something like the edition number of a book would be important to show, but this could be present besides `creators`, `title` information (and thus `description` could not be used in the mutually exclusive setting). I could imagine to have additionally a field for other disambiguation info. This could then replace the `description` field but also be used additionally.
